service/volume.py

```diff
@@ -67,9 +103,47 @@ def create_volume(esh_driver, identity_uuid, name, size,
         metadata=metadata,
         snapshot=snapshot,
         image=image)
+
+    if not success and raise_exception:
+        raise exceptions.VolumeError("The volume failed to be created.")
+
     return success, esh_volume
 
 
+def destroy_volume_or_fail(volume, user, cascade=False):
+    """
+    Destroy the volume specified
+
+    :param cascade: Cascades through and destroy volume snapshots
+                    (defaults is False)
+    :type cascade: ``bool``
+    """
+    provider = volume.instance_source.provider
+    identity = volume.instance_source.created_by_identity
+    # FIXME: fix prepare_driver to take a user directly
+    request = Request(user)
+    driver = prepare_driver(request, provider.uuid, identity.uuid,
+                            raise_exception=True)
+
+    # retrieve volume or fail with not found
+    esh_volume = driver.get_volume(volume.identifier)
+
+    if esh_volume is None:
+        raise exceptions.NotFound(
+            "The `%s` could not be found."
+            % volume.identifier)
+
+    # if cascade True and snapshots exist delete all snapshots
+    if cascade:
+        snapshots = esh_volume.list_snapshots()
+        for snapshot in snapshots:
+            driver.destroy_snapshot(snapshot)
+
+    # destroy the volume successfully or raise an exception
+    if not driver.destroy_volume(esh_volume):
+        raise Exception("Encountered an error destroying the volume.")
```

Comment:
Unsure if generic exception belongs here? Perhaps a CloudService/APIService exception instead to give users a little more guidance in the status code & message?